source/distributed/cell_weights.cc

```diff
@@ -0,0 +1,181 @@
+      std::function<unsigned int(const FiniteElement<dim, spacedim> &)>
+        custom_function)
+    {
+      weighting_function = custom_function;
+    }
+
+
+
+    template <int dim, int spacedim>
+    unsigned int
+    CellWeights<dim, spacedim>::weight_callback(
+      const typename Triangulation<dim, spacedim>::cell_iterator &cell_,
+      const typename Triangulation<dim, spacedim>::CellStatus     status)
+    {
+      // Check if we are still working with the correct combination of
+      // Triangulation and DoFHandler.
+      Assert(&(*triangulation) == &(dof_handler->get_triangulation()),
+             ExcMessage(
+               "Triangulation associated with the DoFHandler has changed!"));
+
+      // Convert cell type from Triangulation to DoFHandler to be able
+      // to access the information about the degrees of freedom.
+      typename dealii::hp::DoFHandler<dim, spacedim>::cell_iterator cell(
+        *cell_, dof_handler);
+
+      // Determine which FiniteElement object will be present on this cell after
+      // refinement and will thus specify the number of degrees of freedom.
+      unsigned int fe_index = numbers::invalid_unsigned_int;
+      switch (status)
+        {
+          case parallel::distributed::Triangulation<dim,
+                                                    spacedim>::CELL_PERSIST:
+          case parallel::distributed::Triangulation<dim, spacedim>::CELL_REFINE:
+          case parallel::distributed::Triangulation<dim,
+                                                    spacedim>::CELL_INVALID:
+            fe_index = cell->active_fe_index();
+            break;
+
+          case parallel::distributed::Triangulation<dim,
+                                                    spacedim>::CELL_COARSEN:
+            {
+              std::set<unsigned int> fe_indices_children;
+              for (unsigned int child_index = 0;
+                   child_index < GeometryInfo<dim>::max_children_per_cell;
+                   ++child_index)
+                {
+                  typename dealii::hp::DoFHandler<dim, spacedim>::cell_iterator
+                    child = cell->child(child_index);
+
+                  fe_indices_children.insert(child->active_fe_index());
+                }
+
+              fe_index = dof_handler->get_fe().find_least_face_dominating_fe(
+                fe_indices_children);
+
+              Assert(fe_index != numbers::invalid_unsigned_int,
+                     ExcMessage(
+                       "No FiniteElement has been found in your FECollection "
+                       "that dominates all children of a cell you are trying "
+                       "to coarsen!"));
+            }
+            break;
+
+          default:
+            Assert(false, ExcInternalError());
+            break;
+        }
+
+      // Return the cell weight determined by the function of choice.
+      return weighting_function(dof_handler->get_fe(fe_index));
```

Comment:
ah, this might require a bit more work if you give the weighting function a cell. but I think it's still worth it. you just have to remember which is the cell with the most dominating fe

Reply:
agreed! i guess I'll do that by introducing a `map` in addition to the `set`

Reply:
no, that won't work. see my comment above. I'll provide the active FE and the cell as parameters now.